pkg/spoke/addon/registration_controller.go

```diff
@@ -0,0 +1,232 @@
+	if errors.IsNotFound(err) {
+		// addon is deleted
+		return c.cleanup(addOnName, ctx)
+	}
+	if err != nil {
+		return err
+	}
+
+	// addon is deleting
+	if !addOn.DeletionTimestamp.IsZero() {
+		return c.cleanup(addOnName, ctx)
+	}
+
+	cachedConfigs := c.addOnRegistrationConfigs[addOnName]
+	configs, err := getRegistrationConfigs(addOnName, addOn.Annotations)
+	if err != nil {
+		return err
+	}
+	staleConfigs, newConfigs := getConfigsDifference(cachedConfigs, configs)
+
+	// stop registraton for the stale registration configs
+	errs := []error{}
+	for _, config := range staleConfigs {
+		if err := c.stopRegistration(config, ctx); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if err := operatorhelpers.NewMultiLineAggregate(errs); err != nil {
+		return err
+	}
+
+	// start registration for the new added configs
+	for _, config := range newConfigs {
+		c.startRegistration(config, ctx)
+	}
+
+	c.addOnRegistrationConfigs[addOnName] = configs
+	return nil
+}
+
+// startRegistration starts a client certificate controller with the given config
+func (c *addOnRegistrationController) startRegistration(config registrationConfig, ctx context.Context) {
+	ctx, stopFunc := context.WithCancel(ctx)
+	kubeInformerFactory := informers.NewSharedInformerFactoryWithOptions(c.kubeClient, 10*time.Minute, informers.WithNamespace(config.InstallationNamespace))
+
+	// build and start a client cert controller
+	clientCertOption := clientcert.ClientCertOption{
+		SecretNamespace: config.InstallationNamespace,
+		SecretName:      config.secretName,
+		AdditonalSecretData: map[string][]byte{
+			clientcert.KubeconfigFile: c.kubeconfigData,
+		},
+	}
+
+	csrOption := clientcert.CSROption{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("addon-%s-%s-", c.clusterName, config.AddOnName),
+			Labels: map[string]string{
+				// the labels are only hints. Anyone could set/modify them.
+				clientcert.ClusterNameLabel: c.clusterName,
+				clientcert.AddonNameLabel:   config.AddOnName,
+			},
+		},
+		Subject:         config.x509Subject(c.clusterName, c.agentName),
+		SignerName:      config.SignerName,
+		EventFilterFunc: createCSREventFilterFunc(c.clusterName, config.AddOnName, config.SignerName),
+	}
+
+	clientCertController := clientcert.NewClientCertificateController(
+		clientCertOption,
+		csrOption,
+		c.hubCSRInformer,
+		c.hubCSRClient,
+		kubeInformerFactory.Core().V1().Secrets(),
+		c.kubeClient.CoreV1(),
+		c.recorder,
+		fmt.Sprintf("ClientCertController@addon:%s:signer:%s", config.AddOnName, config.SignerName),
+	)
+
+	go kubeInformerFactory.Start(ctx.Done())
+	go clientCertController.Run(ctx, 1)
+
+	c.stopRegistrationFuncs[config.hash] = stopFunc
+}
+
+// stopRegistration stops the client certificate controller for the given config
+func (c *addOnRegistrationController) stopRegistration(config registrationConfig, ctx context.Context) error {
+	if stopFunc, ok := c.stopRegistrationFuncs[config.hash]; ok {
```

Comment:
why is this registration function keyed by the config.hash instead of being keyed b y addonName?  It looks like this controller determines the config and the hashed value based on the addonname.  Having a single key would make a litlt easier to read.

Reply:
An addOn may have multiple registrationConfigs with different signers. Each registrationConfig will have a clientcert.NewClientCertificateController started.